File: pythonApi/fantasyCaller.py
import json
import random
import re
import time
import urllib.parse
from dataclasses import dataclass
from html.parser import HTMLParser
from typing import Optional
from urllib.request import Request, urlopen
from urllib.request import HTTPCookieProcessor, build_opener

TARGET = "https://fantasy-frc.cfh.sh"


@dataclass
class User:
    username: str
    password: str
    client: object
    uuid: str = ""
    is_owner: bool = False
    persona: str = ""


@dataclass
class Draft:
    id: int


def load_user_config(path: str) -> str:
    with open(path, "r") as f:
        return f.read()


def parse_users(user_json: str) -> list[User]:
    users_data = json.loads(user_json)
    return [User(**u) for u in users_data]


def create_user(username: str, password: str, persona: str) -> User:
    jar = HTTPCookieProcessor()
    client = build_opener(jar)
    return User(username=username, password=password, client=client, persona=persona)


def init_users(config_path: str) -> list[User]:
    user_json = load_user_config(config_path)
    users = parse_users(user_json)

    for i, user in enumerate(users):
        users[i] = create_user(user.username, user.password, user.persona)

    populate_auth_toks(users)
    print("Starting to make picks")
    return users


def populate_auth_toks(users: list[User]) -> None:
    for user in users:
        print(f"Making login request for user: {user.username}")
        data = urllib.parse.urlencode({"username": user.username, "password": user.password}).encode()
        req = Request(f"{TARGET}/login", data=data, method="POST")
        req.add_header("Content-Type", "application/x-www-form-urlencoded")

        resp = user.client.open(req)
        if resp.getcode() != 200:
            print(f"Failed to login: {user.username}")
            raise Exception("failed to login")

        print(f"Populate auth token request made: {user.username}, Status: {resp.getcode()}")


def create_random_string(min_len: int, max_len: int) -> str:
    alphabet = "abcdefghijklmnopqrstuvwxyz"
    length = random.randint(min_len, max_len)
    return "".join(random.choice(alphabet) for _ in range(length))


def get_player_uuid(owner: User, draft_id: int, username: str) -> str:
    data = urllib.parse.urlencode({
        "description": "",
        "interval": "0",
        "startTime": "0001-01-01T00:00",
        "endTime": "0001-01-01T00:00",
        "draftName": "",
        "search": username
    }).encode()

    req = Request(f"{TARGET}/u/searchPlayers", data=data, method="POST")
    req.add_header("Content-Type", "application/x-www-form-urlencoded")
    req.add_header("Hx-Current-Url", f"{TARGET}/u/draft/{draft_id}/profile")

    resp = owner.client.open(req)
    if resp.getcode() != 200:
        print(f"Failed to search for username: {username}")
        raise Exception("failed to create draft")

    body = resp.read().decode()
    prefix = '<button hx-target="#inviteTable" hx-swap="outerHTML" name="userUuid" value="'
    
    if body.count(prefix) != 1:
        print(f"Did not find only one user: {username}, Draft Id: {draft_id}")
        raise Exception("err: did not find only one user")

    idx = body.index(prefix) + len(prefix)
    sliced = body[idx:]
    uuid_str = sliced[:sliced.index('"')]

    print(f"Found UUID: {username}, UUID: {uuid_str}")
    return uuid_str


def invite_players_to_draft(owner: User, users: list[User], draft: Draft) -> None:
    for user in users:
        if user.username == owner.username:
            continue

        uuid = get_player_uuid(owner, draft.id, user.username)

        data = urllib.parse.urlencode({
            "description": create_random_string(10, 1000),
            "interval": "0",
            "startTime": "0001-01-01T00:00",
            "endTime": "0001-01-01T00:00",
            "draftName": create_random_string(5, 50),
            "search": "",
            "userUuid": uuid
        }).encode()

        req = Request(f"{TARGET}/u/draft/{draft.id}/invitePlayer", data=data, method="POST")
        req.add_header("Content-Type", "application/x-www-form-urlencoded")

        resp = owner.client.open(req)
        if resp.getcode() != 200:
            print(f"Failed to invite player to draft: {user.username}")
            raise Exception("failed to create draft")

        print(f"Request made: {user.username}, Status: {resp.getcode()}")


def create_draft(user: User) -> Draft:
    print(f"Making request to make draft: {user.username}")

    start_time = time.time() + 60
    start_time_str = time.strftime("%Y-%m-%dT%H:%M", time.localtime(start_time))
    end_time_str = time.strftime("%Y-%m-%dT%H:%M", time.localtime(start_time + 600))

    data = urllib.parse.urlencode({
        "description": create_random_string(10, 1000),
        "interval": "0",
        "startTime": start_time_str,
        "endTime": end_time_str,
        "draftName": create_random_string(5, 50)
    }).encode()

    req = Request(f"{TARGET}/u/createDraft", data=data, method="POST")
    req.add_header("Content-Type", "application/x-www-form-urlencoded")

    resp = user.client.open(req)
    if resp.getcode() != 200:
        print(f"Failed to create draft: {user.username}")
        raise Exception("failed to create draft")

    print(f"Create Draft Request Made: {user.username}, Status: {resp.getcode()}")

    redirect = resp.headers.get("Hx-Redirect", "")
    draft_id_str = redirect.split("/")[3]
    draft_id = int(draft_id_str)

    print(f"Created Draft: Id {draft_id}")
    return Draft(id=draft_id)


def get_invite_id(body: str) -> tuple[int, bool]:
    prefix = '<button hx-target="#pendingTable" hx-swap="outerHTML" name="inviteId" value="'
    if body.count(prefix) == 0:
        return -1, False

    idx = body.index(prefix) + len(prefix)
    sliced = body[idx:]
    end_idx = sliced.index('"')
    id_val = int(sliced[:end_idx])

    return id_val, True


def send_accept_invite(user: User, invite_id: int) -> str:
    data = urllib.parse.urlencode({"inviteId": str(invite_id)}).encode()

    req = Request(f"{TARGET}/u/acceptInvite", data=data, method="POST")
    req.add_header("Content-Type", "application/x-www-form-urlencoded")

    resp = user.client.open(req)
    if resp.getcode() != 200:
        raise Exception("failed to accept invite")

    return resp.read().decode()


def accept_invite(user: User) -> None:
    req = Request(f"{TARGET}/u/viewInvites", method="GET")
    resp = user.client.open(req)

    body = resp.read().decode()

    accept_resp_body = ""
    id_val, found = get_invite_id(body)
    r = 0
    while True:
        if found:
            print(f"Sending accept invite request: {user.username}, Id: {id_val}")
            accept_resp_body = send_accept_invite(user, id_val)
            break
        elif r > 5:
            if user.is_owner:
                break
            print(f"Did not find at least one invite id: {username}")
            raise Exception("error: did not find at least one invite id")
        else:
            r += 1
            time.sleep(0.5)
            id_val, found = get_invite_id(body)

    while found:
        id_val, found = get_invite_id(accept_resp_body)
        if found:
            accept_resp_body = send_accept_invite(user, id_val)


def get_draft_id_page(user: User, draft_id: int) -> str:
    req = Request(f"{TARGET}/u/draft/{draft_id}/profile", method="GET")
    resp = user.client.open(req)
    return resp.read().decode()


def get_current_draft_status(user: User, draft_id: int) -> str:
    profile_page = get_draft_id_page(user, draft_id)
    return parse_draft_status(profile_page)


def parse_draft_status(profile_page: str) -> str:
    prefix = 'id="draftStatus">'
    idx = profile_page.find(prefix)
    if idx == -1:
        return ""
    idx += len(prefix)
    end_idx = profile_page[idx:].find("</div>")
    if end_idx == -1:
        return ""
    return profile_page[idx:idx+end_idx].strip()


def wait_until_draft_state(user: User, draft_id: int, requested_status: str, timeout: int) -> None:
    wait_time = 30
    timeout_time = time.time() + timeout
    current_status = get_current_draft_status(user, draft_id)
    
    while current_status != requested_status:
        print(f"Checking if current draft is in requested status: Requested Status: {requested_status}, Current Status: {current_status}")
        if time.time() > timeout_time:
            raise Exception("wait until draft state timeout reached")

        current_status = get_current_draft_status(user, draft_id)
        time.sleep(wait_time)


def start_draft(user: User, draft_id: int) -> None:
    print(f"Start Draft: Draft Id: {draft_id}, User: {user.username}")
    data = urllib.parse.urlencode({}).encode()

    req = Request(f"{TARGET}/u/draft/{draft_id}/startDraft", data=data, method="POST")
    req.add_header("Content-Type", "application/x-www-form-urlencoded")

    resp = user.client.open(req)
    if resp.getcode() != 200:
        print(f"Failed to start draft: {user.username}")
        raise Exception("failed to create draft")

    print(f"Start Draft Request Made: {user.username}, Status: {resp.getcode()}")
    print(f"Started Draft: {draft_id}")


def init_draft(users: list[User]) -> tuple[User, Draft]:
    owner = random.choice(users)
    owner.is_owner = True
    
    draft = create_draft(owner)
    invite_players_to_draft(owner, users, draft)

    for user in users:
        accept_invite(user)

    start_draft(owner, draft.id)

    current_draft_status = get_current_draft_status(owner, draft.id)
    if get_current_draft_status(owner, draft.id) != "Waiting to Start":
        print(f"Got unexpected draft status: Expected 'Waiting to Start', Actual: {current_draft_status}")
        raise Exception("draft status is not correct")

    wait_until_draft_state(owner, draft.id, "Picking", 300)
    return owner, draft


def is_picking_player(user: User, draft_id: int) -> bool:
    print(f"Getting picking player: Draft Id: {draft_id}, User: {user.username}")

    req = Request(f"{TARGET}/u/draft/{draft_id}/pick", method="GET")
    resp = user.client.open(req)

    if resp.getcode() != 200:
        print(f"Failed to get pick page: Draft Id: {draft_id}, User: {user.username}")
        raise Exception("failed to get pick page")

    print(f"Make pick request make: Draft Id: {draft_id}, User: {user.username}, Status: {resp.getcode()}")

    body = resp.read().decode()
    return 'name="pickInput"' in body


def make_pick_request(draft_id: int, user: User, team: int) -> tuple[bool, str]:
    print(f"Make Pick: Draft Id: {draft_id}, User: {user.username}, Team: {team}")
    data = urllib.parse.urlencode({"pickInput": str(team)}).encode()

    req = Request(f"{TARGET}/u/draft/{draft_id}/makePick", data=data, method="POST")
    req.add_header("Content-Type", "application/x-www-form-urlencoded")

    resp = user.client.open(req)

    if resp.getcode() != 200:
        print(f"Failed to make pick: Team: {team}, Draft Id: {draft_id}, User: {user.username}")
        raise Exception("failed to make pick")

    print(f"Make pick request made: Draft Id: {draft_id}, User: {user.username}, Status: {resp.getcode()}")

    body = resp.read().decode()
    has_error = 'id="pickError"' in body

    if has_error:
        class PickErrorParser(HTMLParser):
            def __init__(self):
                super().__init__()
                self.error_text = ""
                self.in_error = False
            
            def handle_starttag(self, tag, attrs):
                if tag == "div":
                    for attr in attrs:
                        if attr[0] == "id" and attr[1] == "pickError":
                            self.in_error = True

            def handle_endtag(self, tag):
                if tag == "div" and self.in_error:
                    self.in_error = False

            def handle_data(self, data):
                if self.in_error:
                    self.error_text += data

        parser = PickErrorParser()
        parser.feed(body)
        return False, parser.error_text.strip()

    return True, ""


def load_valid_teams() -> list[int]:
    with open("./frc-worlds-2025.csv", "r") as f:
        return [int(line.strip()) for line in f]


def main():
    import sys
    if len(sys.argv) != 2:
        print("Usage: python fantasy_caller.py <config_path>")
        sys.exit(1)
    
    config_path = sys.argv[1]
    users = init_users(config_path)
    owner, draft = init_draft(users)

    valid_teams = load_valid_teams()
    print("Users initialized and draft created successfully")


if __name__ == "__main__":
    main()

File: draftAgent/fantasyCaller.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"golang.org/x/net/html"
	"io"
	"log/slog"
	"math/rand/v2"
	"net/http"
	"net/http/cookiejar"
	"net/url"
	"os"
	"reflect"
	"strconv"
	"strings"
	"time"

	"github.com/google/uuid"
)

type User struct {
	Username string `json:"Username"`
	Password string `json:"Password"`
	Client   http.Client
	Uuid     string
	IsOwner  bool
	Persona  DrafterPersona `json:"DraftPersona"`
}

type Draft struct {
	Id int
}

const (
	target = "https://fantasy-frc.cfh.sh"
)

func loadUserConfig(path string) (string, error) {
	file, err := os.ReadFile(path)
	return string(file), err
}

func parseUsers(userJSON string) ([]*User, error) {
	var users []*User
	err := json.Unmarshal([]byte(userJSON), &users)
	return users, err
}

func initUsers(configPath string) ([]*User, error) {
	// Map Username to user struct

	userJSON, err := loadUserConfig(configPath)
	if err != nil {
		return nil, err
	}

	users, err := parseUsers(userJSON)
	if err != nil {
		return nil, err
	}

	for i, user := range users {
		users[i] = createUser(user.Username, user.Password, user.Persona)
	}
	populateAuthToks(users)

	slog.Info("Starting to make picks")

	// Make sure the draft goes to teams playing status
	return users, nil
}

func initDraft(users []*User) (*User, Draft) {
	//Choose a user and create a draft
	owner := users[rand.IntN(len(users))]
	owner.IsOwner = true
	draft := createDraft(owner)
	invitePlayersToDraft(owner, users, draft)
	for _, user := range users {
		acceptInvite(user)
	}

	//Start Draft
	startDraft(owner, draft.Id)

	//Check that draft is in the correct status
	currentDraftStatus := getCurrentDraftStatus(owner, draft.Id)
	if getCurrentDraftStatus(owner, draft.Id) != "Waiting to Start" {
		slog.Error("Got unexpected draft status", "Expected", "Waiting to Start", "Actual", currentDraftStatus)
		panic("draft status is not correct")
	}

	// Wait for draft start time to hit and make sure draft goes into picking
	waitUntilDraftState(owner, draft.Id, "Picking", 300*time.Second)
	return owner, draft
}

func loadValidTeams() []int {
	file, err := os.Open("./frc-worlds-2025.csv")
	if err != nil {
		panic(err)
	}

	var validTeams []int

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		teamNum, err := strconv.Atoi(line)
		if err != nil {
			panic(err)
		}

		validTeams = append(validTeams, teamNum)
	}
	return validTeams
}

func isPickingPlayer(user *User, draftId int) bool {
	slog.Info("Getting picking player", "Draft Id", draftId, "User", user.Username)

	resp, err := user.Client.Get(fmt.Sprintf("%s/u/draft/%d/pick", target, draftId))
	if err != nil {
		slog.Error("Failed to get pick page", "Draft Id", draftId, "Username", user.Username, "Error", err)
		panic(err)
	}

	if resp.StatusCode != 200 {
		slog.Error("Failed to get pick page", "Draft Id", draftId, "User", user.Username)
		panic("failed to get pick page")
	}

	slog.Info("Make pick request make", "Draft Id", draftId, "User", user.Username, "Status", resp.StatusCode)

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		panic("failed to read response body after attempting to get pick page")
	}
	defer resp.Body.Close()

	return strings.Contains(string(body), `name="pickInput"`) && !strings.Contains(string(body), `disabled name="pickInput"`)
}

type CurrentPicks struct {
	PlayerName string   `json:"playerName"`
	Picks      []string `json:"picks"`
}

func getCurrentDraftPicks(user *User, draftId int) ([]CurrentPicks, error) {
	resp, err := user.Client.Get(fmt.Sprintf("%s/u/draft/%d/pick", target, draftId))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	return parseCurrentDraftPicks(string(body)), nil
}

func parseCurrentDraftPicks(htmlContent string) []CurrentPicks {
	doc, err := html.Parse(strings.NewReader(htmlContent))
	if err != nil {
		return nil
	}

	pickTable := findNodeById(doc, "pickTable")
	if pickTable == nil {
		return nil
	}

	var allPicks []CurrentPicks
	cards := findNodesByClass(pickTable, "border", "border-slate-600", "rounded-lg")
	for _, card := range cards {
		nameNode := findNodeByTagAndClass(card, "h3", "text-sm", "font-semibold", "text-slate-200", "text-center", "truncate")
		if nameNode == nil {
			continue
		}
		playerName := extractText(nameNode)

		picksDiv := findNodeByClass(card, "p-3", "space-y-2", "bg-slate-800/50", "rounded-b-lg")
		if picksDiv == nil {
			continue
		}

		var picks []string
		for n := picksDiv.FirstChild; n != nil; n = n.NextSibling {
			if n.Type != html.ElementNode {
				continue
			}

			class := getAttribute(n, "class")
			if class == "" {
				continue
			}

			if containsAll(class, "bg-blue-600") {
				picks = append(picks, extractText(n))
			} else if containsAll(class, "bg-red-600") {
				picks = append(picks, "Skipped")
			}
		}

		allPicks = append(allPicks, CurrentPicks{
			PlayerName: playerName,
			Picks:      picks,
		})
	}

	return allPicks
}

func findNodeById(n *html.Node, id string) *html.Node {
	if n.Type == html.ElementNode {
		for _, attr := range n.Attr {
			if attr.Key == "id" && attr.Val == id {
				return n
			}
		}
	}
	for c := n.FirstChild; c != nil; c = c.NextSibling {
		if found := findNodeById(c, id); found != nil {
			return found
		}
	}
	return nil
}

func findNodesByClass(n *html.Node, classParts ...string) []*html.Node {
	var results []*html.Node
	if n.Type == html.ElementNode {
		class := getAttribute(n, "class")
		if class != "" && containsAll(class, classParts...) {
			results = append(results, n)
		}
	}
	for c := n.FirstChild; c != nil; c = c.NextSibling {
		results = append(results, findNodesByClass(c, classParts...)...)
	}
	return results
}

func findNodeByClass(n *html.Node, classParts ...string) *html.Node {
	if n.Type == html.ElementNode {
		class := getAttribute(n, "class")
		if class != "" && containsAll(class, classParts...) {
			return n
		}
	}
	for c := n.FirstChild; c != nil; c = c.NextSibling {
		if found := findNodeByClass(c, classParts...); found != nil {
			return found
		}
	}
	return nil
}

func findNodeByTagAndClass(n *html.Node, tag string, classParts ...string) *html.Node {
	if n.Type == html.ElementNode && n.Data == tag {
		class := getAttribute(n, "class")
		if class != "" && containsAll(class, classParts...) {
			return n
		}
	}
	for c := n.FirstChild; c != nil; c = c.NextSibling {
		if found := findNodeByTagAndClass(c, tag, classParts...); found != nil {
			return found
		}
	}
	return nil
}

func getAttribute(n *html.Node, key string) string {
	for _, attr := range n.Attr {
		if attr.Key == key {
			return attr.Val
		}
	}
	return ""
}

func extractText(n *html.Node) string {
	if n.Type == html.TextNode {
		return n.Data
	}
	var sb strings.Builder
	for c := n.FirstChild; c != nil; c = c.NextSibling {
		sb.WriteString(extractText(c))
	}
	return sb.String()
}

func containsAll(s string, parts ...string) bool {
	for _, part := range parts {
		if !strings.Contains(s, part) {
			return false
		}
	}
	return true
}

// Returns (success, errorMessage) - errorMessage is empty if pick was made successfully
func makePickRequest(draftId int, user *User, team int) (bool, string) {
	slog.Info("Make Pick", "Draft Id", draftId, "User", user.Username, "Team", team)
	form := url.Values{}
	form.Add("pickInput", strconv.Itoa(team))

	resp, err := user.Client.Post(fmt.Sprintf("%s/u/draft/%d/makePick", target, draftId), "application/x-www-form-urlencoded", strings.NewReader(form.Encode()))
	if err != nil {
		slog.Error("Failed to make pick", "Team", team, "Draft Id", draftId, "Username", user.Username, "Error", err)
		panic(err)
	}

	if resp.StatusCode != 200 {
		slog.Error("Failed to make pick", "Team", team, "Draft Id", draftId, "User", user.Username)
		panic("failed to make pick")
	}

	slog.Info("Make pick request make", "Draft Id", draftId, "User", user.Username, "Status", resp.StatusCode)

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		panic("failed to read response body after attempting to make pick")
	}
	defer resp.Body.Close()

	bodyStr := string(body)
	hasError := strings.Contains(bodyStr, "id=\"pickError\"")

	if hasError {
		doc, err := html.Parse(strings.NewReader(bodyStr))
		if err != nil {
			return false, ""
		}
		errorDiv := findNodeById(doc, "pickError")
		if errorDiv != nil {
			return false, extractText(errorDiv)
		}
		return false, ""
	}

	return true, ""
}

func selectRandomPlayer(users map[string]*User) *User {
	keys := reflect.ValueOf(users).MapKeys()
	return users[keys[rand.IntN(len(keys))].String()]
}

// This will block until the draft is in the desired state or the timeout is hit. Timeout is in milliseconds
func waitUntilDraftState(user *User, draftId int, requestedStatus string, timeout time.Duration) {
	waitTime := 30 * time.Second
	timeoutTime := time.Now().Add(timeout)
	currentStatus := getCurrentDraftStatus(user, draftId)
	for currentStatus != requestedStatus {
		slog.Info("Checking if current draft is in requested status", "Requested Status", requestedStatus, "Current Status", currentStatus)
		if time.Now().After(timeoutTime) {
			panic("wait until draft state timeout reached")
		}

		currentStatus = getCurrentDraftStatus(user, draftId)
		time.Sleep(waitTime)
	}
}

func startDraft(user *User, draftId int) {
	slog.Info("Start Draft", "Draft Id", draftId, "User", user.Username)
	form := url.Values{}

	resp, err := user.Client.Post(fmt.Sprintf("%s/u/draft/%d/startDraft", target, draftId), "application/x-www-form-urlencoded", strings.NewReader(form.Encode()))
	if err != nil {
		slog.Error("Failed to start draft", "Draft Id", draftId, "Username", user.Username, "Error", err)
		panic(err)
	}

	if resp.StatusCode != 200 {
		slog.Error("Failed to start draft", "Draft Id", draftId, "User", user.Username)
		panic("failed to create draft")
	}

	//body, err := io.ReadAll(resp.Body)
	slog.Info("Start Draft Request Made", "Draft Id", draftId, "User", user.Username, "Status", resp.StatusCode)

	slog.Info("Started Draft", "Draft Id", draftId)
}

func acceptInvite(user *User) {
	//Find the draft id, accept the invite, repeat until we dont find any more draft ids
	req, err := http.NewRequest("GET", fmt.Sprintf("%s/u/viewInvites", target), nil)
	if err != nil {
		panic(err)
	}

	resp, err := user.Client.Do(req)
	if err != nil {
		panic(err)
	}

	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		panic(err)
	}

	var acceptRespBody string
	id, found := getInviteId(string(body))
	r := 0
	for {
		if found {
			slog.Info("Sending accept invite request", "User", user.Username, "Id", id)
			acceptRespBody = sendAcceptInvite(user, id)
			break
		} else if r > 5 {
			if user.IsOwner {
				break
			}
			slog.Error("Did not find at least one invite id", "Username", user.Username)
			panic("error: did not find at least one invite id")
		} else {
			r++
			time.Sleep(500 * time.Millisecond)
			id, found = getInviteId(string(body))
		}
	}

	for found {
		id, found = getInviteId(acceptRespBody)
		if found {
			acceptRespBody = sendAcceptInvite(user, id)
		}
	}
}

func getDraftIdPage(user *User, draftId int) string {
	//Find the draft id, accept the invite, repeat until we dont find any more draft ids
	req, err := http.NewRequest("GET", fmt.Sprintf("%s/u/draft/%d/profile", target, draftId), nil)
	if err != nil {
		panic(err)
	}

	resp, err := user.Client.Do(req)
	if err != nil {
		panic(err)
	}

	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		panic(err)
	}

	return string(body)
}

func getCurrentDraftStatus(user *User, draftId int) string {
	//Get Draft Profile package
	profilePage := getDraftIdPage(user, draftId)

	//Parse out the status
	return parseDraftStatus(profilePage)
}

func parseDraftStatus(profilePage string) string {
	prefix := "id=\"draftStatus\">"
	idx := strings.Index(profilePage, prefix)
	if idx == -1 {
		return ""
	}
	idx += len(prefix)
	endIdx := strings.Index(profilePage[idx:], "</div>")
	if endIdx == -1 {
		return ""
	}
	status := strings.TrimSpace(profilePage[idx : idx+endIdx])
	return status
}

func sendAcceptInvite(user *User, inviteId int) string {
	//This should return the respose of the accept request
	form := url.Values{}
	form.Add("inviteId", strconv.Itoa(inviteId))

	req, err := http.NewRequest("POST", fmt.Sprintf("%s/u/acceptInvite", target), strings.NewReader(form.Encode()))
	if err != nil {
		panic(err)
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := user.Client.Do(req)
	if err != nil {
		panic(err)
	}

	if resp.StatusCode != 200 {
		panic("failed to accept invite")
	}

	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		panic(err)
	}

	return string(body)
}

func getInviteId(body string) (int, bool) {
	prefix := "<button hx-target=\"#pendingTable\" hx-swap=\"outerHTML\" name=\"inviteId\" value=\""
	if strings.Count(body, prefix) == 0 {
		return -1, false
	}

	idx := strings.Index(string(body), prefix) + len(prefix)
	sliced := string(body)[idx:]
	endIdx := strings.Index(sliced, "\"")
	id, err := strconv.Atoi(sliced[:endIdx])

	if err != nil {
		panic(err)
	}

	return id, true
}

func createUser(username string, password string, persona DrafterPersona) *User {
	jar, err := cookiejar.New(nil)
	if err != nil {
		panic(err)
	}
	return &User{
		Username: username,
		Password: password,
		Client: http.Client{
			Jar: jar,
		},
		Persona: persona,
	}
}

func createRandomString(minLen int, maxLen int) string {
	alphabet := "abcdefghijklmnopqrstuvwxyz"
	length := rand.IntN(maxLen-minLen) + minLen
	var sb strings.Builder
	for range length {
		sb.WriteByte(alphabet[rand.IntN(len(alphabet))])
	}
	return sb.String()
}

func getPlayerUUID(owner *User, draftId int, username string) uuid.UUID {
	form := url.Values{}
	form.Add("description", "")
	form.Add("interval", "0")
	form.Add("startTime", "0001-01-01T00:00")
	form.Add("endTime", "0001-01-01T00:00")
	form.Add("draftName", "")
	form.Add("search", username)

	req, err := http.NewRequest("POST", fmt.Sprintf("%s/u/searchPlayers", target), strings.NewReader(form.Encode()))
	if err != nil {
		slog.Error("Failed to create new request", "Error", err)
		panic(err)
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Hx-Current-Url", fmt.Sprintf("%s/u/draft/%d/profile", target, draftId))

	resp, err := owner.Client.Do(req)
	if err != nil {
		slog.Error("Failed to search for player", "Username", username, "Error", err)
		panic(err)
	}

	if resp.StatusCode != 200 {
		slog.Error("Failed to search for username", "User", username)
		panic("failed to create draft")
	}

	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	slog.Info("Request made", "User", username, "Status", resp.StatusCode)

	prefix := "<button hx-target=\"#inviteTable\" hx-swap=\"outerHTML\" name=\"userUuid\" value=\""
	if strings.Count(string(body), prefix) != 1 {
		slog.Error("Did not find only one user", "Username", username, "Draft Id", draftId)
		panic("err: did not find only one user")
	}

	idx := strings.Index(string(body), prefix) + len(prefix)
	sliced := string(body)[idx:]
	uuidStr := sliced[:strings.Index(sliced, "\"")]

	uuid, err := uuid.Parse(uuidStr)

	if err != nil {
		panic(err)
	} else {
		slog.Info("Found UUID", "Username", username, "UUID", uuid)
	}

	return uuid
}

func invitePlayersToDraft(owner *User, users []*User, draft Draft) {
	for _, user := range users {
		if user.Username == owner.Username {
			continue
		}

		form := url.Values{}
		form.Add("description", createRandomString(10, 1000))
		form.Add("interval", "0")
		form.Add("startTime", "0001-01-01T00:00")
		form.Add("endTime", "0001-01-01T00:00")
		form.Add("draftName", createRandomString(5, 50))
		form.Add("search", "")
		form.Add("userUuid", getPlayerUUID(owner, draft.Id, user.Username).String())

		resp, err := owner.Client.Post(fmt.Sprintf("%s/u/draft/%d/invitePlayer", target, draft.Id), "application/x-www-form-urlencoded", strings.NewReader(form.Encode()))
		if err != nil {
			slog.Error("Failed to invite player to draft", "Username", user.Username, "Error", err)
			panic(err)
		}

		if resp.StatusCode != 200 {
			slog.Error("Failed to invite player to draft", "User", user.Username)
			panic("failed to create draft")
		}

		//body, err := io.ReadAll(resp.Body)
		slog.Info("Request made", "User", user.Username, "Status", resp.StatusCode)
	}
}

func createDraft(user *User) Draft {
	slog.Info("Making request to make draft", "User", user.Username)

	startTime := time.Now().Add(1 * time.Minute)

	layout := "2006-01-02T15:04:05"
	form := url.Values{}
	form.Add("description", createRandomString(10, 1000))
	form.Add("interval", "0")
	form.Add("startTime", startTime.Format(layout))
	form.Add("endTime", startTime.Add(10*time.Minute).Format(layout))
	form.Add("draftName", createRandomString(5, 50))

	resp, err := user.Client.Post(target+"/u/createDraft", "application/x-www-form-urlencoded", strings.NewReader(form.Encode()))
	if err != nil {
		slog.Error("Failed to create draft", "Username", user.Username, "Error", err)
		panic(err)
	}

	if resp.StatusCode != 200 {
		slog.Error("Failed to create draft", "User", user.Username)
		panic("failed to create draft")
	}

	//body, err := io.ReadAll(resp.Body)
	slog.Info("Create Draft Request Made", "User", user.Username, "Status", resp.StatusCode)

	// Get Draft Id
	draftIdStr := strings.Split(resp.Header.Get("Hx-Redirect"), "/")[3]
	slog.Info("Parsed draft id string", "Draft Id String", draftIdStr)
	draftId, err := strconv.Atoi(draftIdStr)
	if err != nil {
		slog.Error("Failed to parse draft id from redirect")
		panic(err)
	}

	slog.Info("Created Draft", "Id", draftId)

	return Draft{
		Id: draftId,
	}
}

func populateAuthToks(users []*User) {
	for _, user := range users {
		slog.Info("Making login request", "User", user.Username)
		form := url.Values{}
		form.Add("username", user.Username)
		form.Add("password", user.Password)
		resp, err := user.Client.Post(target+"/login", "application/x-www-form-urlencoded", strings.NewReader(form.Encode()))
		if err != nil {
			slog.Error("Failed login", "Username", user.Username, "Error", err)
			panic(err)
		}

		defer resp.Body.Close()

		if resp.StatusCode != 200 {
			slog.Error("Failed to login", "User", user.Username)
			panic("failed to login")
		}

		//body, err := io.ReadAll(resp.Body)
		slog.Info("Populate auth token request made", "User", user.Username, "Status", resp.StatusCode)
	}
}
